ksql-engine/src/main/java/io/confluent/ksql/function/udf/array/Entries.java

```diff
@@ -0,0 +1,109 @@
+import io.confluent.ksql.function.udf.UdfParameter;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+
+/**
+ * This UDF constructs an array of structs from the entries in a map. Each struct has a field with
+ * name "K" containing the key (this is always a String) and a field with name "V" holding the
+ * value;
+ */
+@UdfDescription(name = "ENTRIES", description = "Construct an array from the entries in a map")
+public class Entries {
+
+  private static final Schema INT_STRUCT_SCHEMA = buildStructSchema(Schema.OPTIONAL_INT32_SCHEMA);
+  private static final Schema BIGINT_STRUCT_SCHEMA = buildStructSchema(
+      Schema.OPTIONAL_INT64_SCHEMA);
+  private static final Schema DOUBLE_STRUCT_SCHEMA = buildStructSchema(
+      Schema.OPTIONAL_FLOAT64_SCHEMA);
+  private static final Schema BOOLEAN_STRUCT_SCHEMA = buildStructSchema(
+      Schema.OPTIONAL_BOOLEAN_SCHEMA);
+  private static final Schema STRING_STRUCT_SCHEMA = buildStructSchema(
+      Schema.OPTIONAL_STRING_SCHEMA);
+  private static final String KEY_FIELD_NAME = "K";
+  private static final String VALUE_FIELD_NAME = "V";
+
+  private static Schema buildStructSchema(final Schema valueSchema) {
+    return SchemaBuilder.struct().field(KEY_FIELD_NAME, Schema.OPTIONAL_STRING_SCHEMA)
+        .field(VALUE_FIELD_NAME, valueSchema).optional().build();
+  }
+
+  @Udf(schema = "ARRAY<STRUCT<K STRING, V INT>>")
+  public List<Struct> entriesInt(
+      @UdfParameter final Map<String, Integer> map, @UdfParameter final boolean sorted
+  ) {
+    return entries(map, INT_STRUCT_SCHEMA, sorted);
+  }
+
+  @Udf(schema = "ARRAY<STRUCT<K STRING, V BIGINT>>")
+  public List<Struct> entriesBigInt(
+      @UdfParameter final Map<String, Long> map, @UdfParameter final boolean sorted
+  ) {
+    return entries(map, BIGINT_STRUCT_SCHEMA, sorted);
+  }
+
+  @Udf(schema = "ARRAY<STRUCT<K STRING, V DOUBLE>>")
+  public List<Struct> entriesDouble(
+      @UdfParameter final Map<String, Double> map, @UdfParameter final boolean sorted
+  ) {
+    return entries(map, DOUBLE_STRUCT_SCHEMA, sorted);
+  }
+
+  @Udf(schema = "ARRAY<STRUCT<K STRING, V BOOLEAN>>")
+  public List<Struct> entriesBoolean(
+      @UdfParameter final Map<String, Boolean> map, @UdfParameter final boolean sorted
+  ) {
+    return entries(map, BOOLEAN_STRUCT_SCHEMA, sorted);
+  }
+
+  @Udf(schema = "ARRAY<STRUCT<K STRING, V STRING>>")
+  public List<Struct> entriesString(
+      @UdfParameter final Map<String, String> map, @UdfParameter final boolean sorted
+  ) {
+    return entries(map, STRING_STRUCT_SCHEMA, sorted);
+  }
+
+  private <T> List<Struct> entries(
+      final Map<String, T> map, final Schema structSchema, @UdfParameter final boolean sorted
```

Comment:
no need for the `@UdfParameter` annotation here!

Reply:
copypasta error ;)

Reply:
done